### PR TITLE
[Engine] clean key uniqueness mode

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This is a follow up of #14175 and #14176.  

Basically
-  we kill the `On` mode which is not used anymore.
- we factorize the visibility check.
- we rename `LocalContractNotVisible` into `ContractNotVisible`

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 1]
```diff
@@ -229,16 +230,19 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
       }
     }
 
-    /**  Must be used to handle lookups iff [[com.daml.lf.transaction.ContractKeyUniquenessMode.byKeyOnly]] is false
+    /**  Must be used to handle lookups iff in [[com.daml.lf.transaction.ContractKeyUniquenessMode.Off]] mode
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:233]:
```suggestion
    /**  Must be used to handle lookups iff in [[com.daml.lf.transaction.ContractKeyUniquenessMode.Strict]] mode
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 2]
```diff
@@ -229,16 +230,19 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
...
-      require(!mode.byKeyOnly, "This method can only be used if all key nodes are considered")
+      if (mode != ContractKeyUniquenessMode.Strict)
+        throw new UnsupportedOperationException(
+          "handleLookup can only be used if all key nodes are considered"
+        )
       visitLookup(lookup.templateId, lookup.key.key, lookup.result, lookup.result).left.map(Left(_))
     }
 
-    /** Must be used to handle lookups iff [[com.daml.lf.transaction.ContractKeyUniquenessMode.byKeyOnly]] is true
+    /** Must be used to handle lookups iff in [[com.daml.lf.transaction.ContractKeyUniquenessMode.Off]] mode
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:245]: Minor: Do you want to throw an `UnsupportedOperationException` in this method analogously to `handleLookup` if we're in the wrong mode?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 3]
```diff
@@ -382,13 +386,13 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
     private def withinRollbackScope: Boolean = rollbackStack.nonEmpty
 
     /** Let `resolver` be a [[KeyResolver]] that can be used during interpretation to obtain a transaction `tx`
-      * in modes [[com.daml.lf.transaction.ContractKeyUniquenessMode.ContractByKeyUniquenessMode]],
+      * in modes [[com.daml.lf.transaction.ContractKeyUniquenessMode.Off]],
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:389]:
```suggestion
      * in mode [[com.daml.lf.transaction.ContractKeyUniquenessMode.Off]],
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 4]
```diff
@@ -525,7 +529,6 @@ object ContractStateMachine {
     *     nodes under a rollback.
     *
     *   - In modes [[com.daml.lf.transaction.ContractKeyUniquenessMode.Off]]
```
Comment (andreaslochbihler-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:531]:
```suggestion
    *   - In mode [[com.daml.lf.transaction.ContractKeyUniquenessMode.Off]]
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/GlobalKey.scala [hunk 5]
```diff
@@ -49,24 +49,13 @@ final case class GlobalKeyWithMaintainers(
...
-}
 object ContractKeyUniquenessMode {
 
-  /** Considers only by-key and create nodes for the contract key uniqueness checks */
-  sealed trait ContractByKeyUniquenessMode extends ContractKeyUniquenessMode {
-    override def byKeyOnly: Boolean = true
-  }
-  case object On extends ContractByKeyUniquenessMode
-
   /** Note that no stable semantics are provided for off mode. */
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/GlobalKey.scala:56]:
```suggestion
  /** Disable key uniqueness checks and only consider byKey operations.
      Note that no stable semantics are provided for off mode. */
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/GlobalKey.scala [hunk 6]
```diff
@@ -49,24 +49,13 @@ final case class GlobalKeyWithMaintainers(
...
-  case object Off extends ContractByKeyUniquenessMode
+  case object Off extends ContractKeyUniquenessMode
 
   /** Considers all nodes mentioning keys for contract key uniqueness */
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/GlobalKey.scala:59]:
```suggestion
  /** Considers all nodes mentioning keys as byKey operations and checks for contract key uniqueness. */
```

**Review by cocreature — APPROVED** [review]
lgtm thanks!

**Pull request merged**